### ISSUE TITLE
Transition children elements

### ISSUE THREAD

**Issue by Elfayer** [vuejs/vue#3477]
### Vue.js version
1.0.26

### Reproduction Link
http://jsfiddle.net/jf6b5st9/

### Steps to reproduce
- Click on the button to open the slider (OK)
- Click on the darkened zone on the left to close the slider (FAIL)
  -> The element is never hide. We can't click on the body anymore.

### What is Expected?
The element should be removed after animation.

### What is actually happening?
The element is not removed from the DOM. (`afterLeave` is never called)

## More details
**Note:** I had to add the CSS `transition: all 2s;` to the class `.slider-transition`. Otherwise VueJS doesn't know about the transition time.

### Improvement
I feel like the way I'm using the transition system is a bit "hacky" although I have no idea how to avoid it.
Ideally I'd like to be able to call transitions as follow:

```
<div id="slider" v-if="opened">
  <div id="slider-backdrop" @click="opened = false" transition="fade"></div>
  <div id="slider-content" transition="slide"></div>
</div>
```

But the `transition` attribute is only allowed at the `v-if` level.

Did I miss something ?
What do you think ?

**Comment by Elfayer** [comment]
Here is a "trick" I found to make it work : http://jsfiddle.net/0gmusue5/1/

The solution is still to add a false transition to the VueJS transitioned element and add a false attribute `opacity: 0.999` to make it work (`opacity: 1` doesn't work, I guess there needs to be a transition in a way). It still fills like a hack to me. Any advice or opinion about this ?

**Comment by simplesmiler** [comment]
Yeah, I think there is no way for Vue to know which nested element to listen for `transitionend` event on, so it is listening only on the element the `transition` is on. Having fake "master transition" is the best workaround I can think of.

**Comment by Elfayer** [comment]
@simplesmiler That's why I asked for an improvement, it is definitely possible for Vue to get/cache the transitions that will happen at an element `v-if` of `v-show`. You just have to get all the elements below that have a `transition` attribute.

**Comment by dmx374** [comment]
Also, it would be really great to add an $update method to arrays, so we can update existing properties or add new ones (and make them reactive).

I've made a quick and dirty patch working for my needs : 

```js

	/**
	 * update an element at the given index with new keys,
	 * and emits corresponding event.
	 *
	 * @param {Number} index
	 * @param {*} keys
	 * @return {*} - replaced element
	 */
	def(arrayProto, '$update', function $update(index, keys) {
		if (index >= this.length) {
			this.length = Number(index) + 1;
		}
		var val = [],
			source = this[index];

		for (var i in source) {
			val[i] = source[i];
		}
		for (var i in keys) {
			val[i] = keys[i];
		}
		return this.splice(index, 1, val)[0];
	});
```

**Comment by LinusBorg** [comment]
@dmx374 What has that to do with the topic of this issue? It seems to be a completely seperate feature request.

If so, please open a seperate issue for this.

**Comment by dmx374** [comment]
I tried, but impossible to create a new issue, the submit button is disabled, I do not know why.

**Comment by LinusBorg** [comment]
I will nonetheless remove your post, as this issue should stay on topic.

**Comment by Elfayer** [comment]
It's a bit easier using Vue2, though it still requires the top level element of the transition to know the max delay.

```
.slider-enter-active, .slider-leave-active {
  transition: all 1s; // trick
}

.slider-enter, .slider-leave-active {
  #slider-backdrop {
    opacity: 0;
  }
  #slider-content {
    transform: translateX(calc(100% + 36px));
  }
}
```

Demo : http://jsfiddle.net/0gmusue5/5/
